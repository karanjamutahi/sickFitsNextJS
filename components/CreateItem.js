import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';

import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import ErrorMessage from './ErrorMessage';
import { ALL_ITEMS_QUERY } from './Items'; //dangerous to refetch query because it could take long

const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION(
        $title: String!
        $description: String!
        $price: Int!
        $image: String
        $largeImage: String
    ) {
        createItem(
        title: $title
        description: $description
        price: $price
        image: $image
        largeImage: $largeImage
        ) {
            id
        }
    }
`;

class CreateItem extends Component {
    state = {
        title: '',
        description: '',
        image: '',
        largeImage: '',
        price: 0
    };

    handleChange = (e) => {
        const { name,type,value } = e.target;
        const val = type === 'number' ? parseFloat(value) : value;
        this.setState({ [name]: val});
    };

    handleImage = async (e) => {

        //1. Upload Image
        const files = e.target.files;
        const data = new FormData();
        data.append('file',files[0]);
        data.append('upload_preset', 'sickfits');
        const res = await fetch('https://api.cloudinary.com/v1_1/karanjamutahi/image/upload', {
            method: "POST",
            body: data
        });
        const file = await res.json(); 
        this.setState({
            image: file.secure_url,
            largeImage: file.eager[0].secure_url,
        })
    };

    render(){
        return(
            <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state} refetchQueries={ALL_ITEMS_QUERY}>
                {(createItem, { loading, error})=>(

            <Form data-test="CreateItem" onSubmit={ async e => {
                // 1.Stop form from submitting
                e.preventDefault();
                // 2. Poll for image to upload & return a URL
                /*
                while(!this.state.largeImage) { 
                    loading=true; //Hold the line & Disable the fieldset
                }
                loading = false; //Re-Enable the fieldset
                */
                //3. Run the mutation
                const { data } = await createItem();
                // 3. Change them to the Single Item Page
                Router.push(`/item/[pid]`, `/item/${data.createItem.id}`);
            }}>
                <h2>Sell an Item</h2>

                <ErrorMessage error={error}></ErrorMessage>
                <fieldset disabled={loading} aria-busy={loading}>

                    {this.state.image && <img src={this.state.image} alt="Image Preview"></img> /**TODO: Add an x button to delete entry from cloudinary and upload a new image */}
                    <label htmlFor="file">
                        Upload Image
                        <input type="file" name="file" id="file" placeholder='Upload Image' required onChange={this.handleImage}/>
                    </label>

                    <label htmlFor="title">
                        Title
                        <input type="text" name="title" id="title" placeholder='Title' required value={this.state.title} onChange={this.handleChange}/>
                    </label>

                    <label htmlFor="price">
                        Price (KSH)
                        <input type="number" name="price" id="price" placeholder='Price' required value={this.state.price} onChange={this.handleChange}/>
                    </label>

                    <label htmlFor="description">
                        Description
                        <textarea type="number" name="description" id="description" placeholder='Enter A Description' required value={this.state.description} onChange={this.handleChange}/>
                    </label>

                    <button type='Submit'>Add Item</button>
                </fieldset>
            </Form>
            )}
        </Mutation>
        );
    }
};

export default CreateItem;
export { CREATE_ITEM_MUTATION };