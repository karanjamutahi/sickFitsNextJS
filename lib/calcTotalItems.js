function totalItems(cart) {
    return cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)
}

export default totalItems;