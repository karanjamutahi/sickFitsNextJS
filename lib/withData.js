import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { devEndpoint, prodEndpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? devEndpoint : devEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            //read cartopen value from cache
             const { cartOpen } = cache.readQuery({
               query: LOCAL_STATE_QUERY,
             });
             //Write the cart State to the opposite
             const data = {
               data: {
                 cartOpen: ! cartOpen
               }
             };
             cache.writeData(data);
             return data; 
          }
        }
      },
      defaults: {
        cartOpen: false,
      }
    }
  });
}

export default withApollo(createClient);
