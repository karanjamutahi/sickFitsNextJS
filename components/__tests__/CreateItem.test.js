import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import Router from 'next/Router';
import { MockedProvider } from 'react-apollo/test-utils';
import CreateItem, { CREATE_ITEM_MUTATION } from '../CreateItem';
import { fakeItem } from '../../lib/testUtils';
import { resultKeyNameFromField } from 'apollo-utilities';

const dogImage = 'https://unsplash.com/dog.jpg';
//Mock the global fetch API
global.fetch = jest.fn().mockResolvedValue({
    json: () => ({
        secure_url: dogImage,
        eager: [{ secure_url: dogImage}]
    })
});

describe("Create Item Component", () => {
    it('Renders and matches snapshot', () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateItem></CreateItem>
            </MockedProvider>
        );

        expect(toJSON(wrapper.find('form[data-test="CreateItem"]'))).toMatchSnapshot();
    });

    it("Uploads a file when changed", async() => {
        const wrapper = mount(
            <MockedProvider>
                <CreateItem></CreateItem>
            </MockedProvider>
        );

        const input = wrapper.find('input[type="file"]');
        //console.log(input.debug());
        input.simulate('change', { target: { files: ['fakedog.jpg']}});
        await wait();
        const component = wrapper.find('CreateItem').instance();
        //console.log(component.state);
        expect(component.state.image).toEqual(dogImage);
        expect(component.state.largeImage).toEqual(dogImage);
        expect(global.fetch).toHaveBeenCalled();
        global.fetch.mockReset();
    });

    it("handles state updating", async () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateItem></CreateItem>
            </MockedProvider>
        );
        const input = wrapper.find('input[name="title"]');
        //console.log(input.debug());
        input.simulate('change', { target:{
            name: 'title',
            value: 'dog',
            type: 'text'
        }});
        await wait();
        const instance = wrapper.find('CreateItem').instance();
        expect(instance.state.title).toEqual('dog');
    });

    it("creates an item when the form is submitted", async () => {
        const item = fakeItem();
        //console.log(item);

        const mocks = [{
            request: { query: CREATE_ITEM_MUTATION, variables: {
                title: item.title,
                description: item.description,
                price: item.price,
                image: '',
                largeImage: '',
            } },
            result: { data: {
                createItem: {
                    ...item,
                    __typename: 'Item'
                }
            }}
        }];

        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <CreateItem></CreateItem>
            </MockedProvider>
        );
        //console.log(wrapper.debug());

        const titleInput = wrapper.find('input[name="title"]');
        //console.log(titleInput.debug());
        titleInput.simulate('change', { target:{
            name: 'title',
            value: item.title,
        }});

        const descriptionInput = wrapper.find('#description');
        descriptionInput.simulate('change', { target: {
            name: 'description',
            value: item.description
        }});

        const priceInput = wrapper.find('input[name="price"]');
        priceInput.simulate('change', { target: {
            name: 'price',
            value: item.price,
            type: 'number'
        }});

        //Mock router
        Router.router = { push: jest.fn()};
        const form = wrapper.find('form');
        //console.log(form.debug());
        form.simulate('submit');    
        await wait(50);
        expect(Router.router.push).toHaveBeenCalled();
        expect(Router.router.push).toHaveBeenCalledWith("/item/[pid]", "/item/abc123");
    });
});