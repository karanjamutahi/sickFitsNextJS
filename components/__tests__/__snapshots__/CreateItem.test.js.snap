// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Item Component Renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 gvKJrt"
  data-test="CreateItem"
  onSubmit={[Function]}
>
  <h2>
    Sell an Item
  </h2>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="file"
    >
      Upload Image
      <input
        id="file"
        name="file"
        onChange={[Function]}
        placeholder="Upload Image"
        required={true}
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="price"
    >
      Price (KSH)
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="description"
    >
      Description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter A Description"
        required={true}
        type="number"
        value=""
      />
    </label>
    <button
      type="Submit"
    >
      Add Item
    </button>
  </fieldset>
</form>
`;
